### PR TITLE
feat: [libp2p] refactor Hello protocol into its own behaviour

### PR DESCRIPTION
**Pull request by hanabi1224**
## Summary of changes

Changes introduced in this pull request:
- 

## Reference issue to close (if applicable)

Closes https://github.com/ChainSafe/forest/issues/2496

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG](https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md) is up-to-date. All user-facing changes should be reflected in this document.

### CODE REVIEW

File: blockchain/chain_sync/src/chain_muxer.rs [hunk 1]
```diff
@@ -271,19 +268,9 @@ where
...
             match response {
-                Some(Ok(_res)) => {
+                Some(_) => {
                     network.peer_manager().log_success(peer_id, dur).await;
                 }
-                Some(Err(why)) => match why {
-                    RequestResponseError::ConnectionClosed
-                    | RequestResponseError::DialFailure
-                    | RequestResponseError::UnsupportedProtocols => {
-                        network.peer_manager().mark_peer_bad(peer_id).await;
```
Comment (elmattic) [inline blockchain/chain_sync/src/chain_muxer.rs:281]: This is not an issue removing this code? Or is `mark_peer_bad` done elsewhere?

File: blockchain/chain_sync/src/network_context.rs [hunk 2]
```diff
@@ -369,18 +362,13 @@ where
                 response_channel: tx,
             })
             .await
-            .map_err(|_| "Failed to send hello request: receiver dropped")?;
+            .map_err(|_| anyhow::anyhow!("Failed to send hello request: receiver dropped"))?;
```
Comment (elmattic) [inline blockchain/chain_sync/src/network_context.rs:365]: You can make use of context here. See https://docs.rs/anyhow/latest/anyhow/trait.Context.html

**Review by elmattic — CHANGES_REQUESTED** [review]
Looks like a good refactor to me.

File: blockchain/chain_sync/src/chain_muxer.rs (hunk 1 above)
Comment (hanabi1224) [reply blockchain/chain_sync/src/chain_muxer.rs:281]: This logic is never hit, in the old code it always sends Ok(response) to the channel: https://github.com/ChainSafe/forest/pull/2498/files#diff-9bebebe2f8a520466a9fe8ca9c212cc919b05b9ca67c44773742cf31f96b1a29L629
```rust
if tx.send(Ok(response)).is_err() {
    warn!("Fail to send Hello response");
}
```

Comment (hanabi1224) [reply blockchain/chain_sync/src/chain_muxer.rs:282]: You are right, I should have added this logic to here, now fixed: https://github.com/ChainSafe/forest/pull/2498/files#diff-9bebebe2f8a520466a9fe8ca9c212cc919b05b9ca67c44773742cf31f96b1a29R627

File: blockchain/chain_sync/src/network_context.rs (hunk 2 above)
Comment (hanabi1224) [reply blockchain/chain_sync/src/network_context.rs]: Good one! I didn't know that, fixed

**Review by elmattic — APPROVED** [review]

**Pull request merged**